File: leetcode/301-400/319-bulb-switcher/bench/bulb_switcher.c
/* Benchmark mirror of LeetCode #319 — the round simulation.
   Same algorithm as bench/bulb_switcher.kara: PASSES passes, each simulating
   n rounds over an n-bulb byte array and folding the count of lit bulbs
   together with the sum of their indices. */
#include <stdio.h>
#include <stdlib.h>

#define BULBS   6000000LL
#define PASSES  10LL
#define STRIDE  90011LL
#define MASKMOD 1073741823LL

int main(void) {
    unsigned char *on = calloc(BULBS + 1, 1);
    if (!on) return 1;

    long long sink = 0;
    for (long long p = 0; p < PASSES; p++) {
        long long n = BULBS - p * STRIDE;

        for (long long b = 0; b <= n; b++) on[b] = 0;

        for (long long step = 1; step <= n; step++)
            for (long long b = step; b <= n; b += step)
                on[b] ^= 1;

        long long count = 0, idx_sum = 0;
        for (long long b = 1; b <= n; b++) {
            if (on[b] == 1) {
                count++;
                idx_sum = (idx_sum + b) % MASKMOD;
            }
        }
        sink = (sink * 31 + count * 7919 + idx_sum) % MASKMOD;
    }

    printf("checksum %lld\n", sink);
    free(on);
    return 0;
}

File: leetcode/301-400/319-bulb-switcher/bench/bulb_switcher.py
"""Benchmark mirror of LeetCode #319 — the round simulation.

Same algorithm as bench/bulb_switcher.kara: PASSES passes, each simulating n
rounds over an n-bulb byte array and folding the count of lit bulbs together
with the sum of their indices."""

BULBS = 6000000
PASSES = 10
STRIDE = 90011
MASKMOD = 1073741823


def main() -> None:
    on = bytearray(BULBS + 1)

    sink = 0
    for p in range(PASSES):
        n = BULBS - p * STRIDE

        for b in range(n + 1):
            on[b] = 0

        for step in range(1, n + 1):
            for b in range(step, n + 1, step):
                on[b] ^= 1

        count = 0
        idx_sum = 0
        for b in range(1, n + 1):
            if on[b] == 1:
                count += 1
                idx_sum = (idx_sum + b) % MASKMOD
        sink = (sink * 31 + count * 7919 + idx_sum) % MASKMOD

    print(f"checksum {sink}")


if __name__ == "__main__":
    main()
